File: Linked List/singly/addition/list.cpp
// Singly Linked List - Addition (C++)
// This program demonstrates how to add (insert) nodes in a singly linked list.
// We'll insert at the head and at the tail, then print the result.

#include <iostream>
using namespace std;

// Node represents an element with data and a pointer to the next node
struct Node {
    int data;
    Node* next;
    Node(int value) : data(value), next(nullptr) {}
};

// insertAtHead adds a new node before the current head and returns the new head
Node* insertAtHead(Node* head, int value) {
    Node* newNode = new Node(value); // create the new node
    newNode->next = head;            // link new node to the old head
    return newNode;                  // new node becomes the new head
}

// insertAtTail adds a new node at the end of the list
void insertAtTail(Node*& head, int value) {
    Node* newNode = new Node(value);
    if (head == nullptr) { // empty list: new node is the head
        head = newNode;
        return;
    }
    Node* current = head;
    while (current->next != nullptr) { // move to the last node
        current = current->next;
    }
    current->next = newNode; // attach new node at the end
}

// traverse prints all elements in the list
void traverse(Node* head) {
    Node* current = head;
    while (current != nullptr) {
        cout << current->data << " ";
        current = current->next;
    }
    cout << "\n";
}

int main() {
    // Start with a small list: 20 -> 30
    Node* head = nullptr;
    insertAtTail(head, 20);
    insertAtTail(head, 30);

    cout << "Initial list: ";
    traverse(head);

    // Add at head: insert 10 at the front -> 10 -> 20 -> 30
    head = insertAtHead(head, 10);
    cout << "After insertAtHead(10): ";
    traverse(head);

    // Add at tail: insert 40 at the end -> 10 -> 20 -> 30 -> 40
    insertAtTail(head, 40);
    cout << "After insertAtTail(40): ";
    traverse(head);

    // Cleanup memory
    Node* cur = head;
    while (cur != nullptr) {
        Node* toDelete = cur;
        cur = cur->next;
        delete toDelete;
    }
    return 0;
}

File: Linked List/singly/traversing/list.py
# Singly Linked List - Traversing (Python)
# This script demonstrates how to traverse a singly linked list.
# A Node stores data and a reference to the next Node.

class Node:
    """Represents a single element in the singly linked list.

    Each node has:
    - data: value stored in this node
    - next: reference to the next node (or None if this is the last node)
    """

    def __init__(self, value: int) -> None:
        self.data = value
        self.next = None


def traverse(head: Node) -> None:
    """Print all node values from head to end.

    We use a temporary pointer that follows next references until None.
    """
    current = head  # start at the first node
    while current is not None:  # stop when we reach the end (None)
        print(current.data, end=" ")
        current = current.next  # follow the link to the next node
    print()


if __name__ == "__main__":
    # Create a small linked list with 3 nodes: 10 -> 20 -> 30 -> None
    head = Node(10)
    head.next = Node(20)
    head.next.next = Node(30)

    # Traverse and print the list
    print("Traversing singly linked list:", end=" ")
    traverse(head)


